### ISSUE TITLE
Feature request: way to limit number of word forms and definitions copied with c+t

### ISSUE THREAD

**Issue by demilade5** [birchill/10ten-ja-reader#1305]
For example: 落ち着く

If you c+t that word, 10ten copies 5 forms and 6 defintions to the clipboard.

Would be nice to have a setting that allows you to specifiy the (max) number of forms and definitions 10ten copies by default.

**Comment by birtles** [comment]
Hi! Thanks for filing this issue. Can I ask what you are using the copy functionality for? Knowing that might help us find a good solution that works for other people too.

For example, perhaps we should make it possible to copy it with line breaks for each definition so that it's easier to drop the later definitions?

For the forms, we could certainly make it just copy the most common form (e.g. if there is only one form with a ★, we could just copy that). In fact, maybe we should do that always.

With those two changes the result would look like:

```
落ち着く
1. (v5k,vi) to calm down; to compose oneself; to regain presence of mind; to relax
2. (v5k,vi) to calm down; to settle down; to die down; to become stable; to abate
3. (v5k,vi) to settle down (in a location, job, etc.); to settle in
4. (v5k,vi) to be settled; to be fixed; to have been reached (of an arrangement, conclusion, etc.)
5. (v5k,vi) to harmonize with; to harmonise with; to match; to suit; to fit
6. (v5k,vi) to be unobtrusive; to be quiet; to be subdued (usu. used pronominally as 落ち着いた)
```

Or we could add a simple mode that only returns the first definition and drops annotations like the part-of-speech? e.g.

```
落ち着く to calm down; to compose oneself; to regain presence of mind; to relax
```

**Comment by demilade5** [comment]
Hi birtles!

In my case, I use the c+t function for quickly pasting and making flash cards in different SRS apps. Sometimes I just want each card to have maybe 2 or 3 definitions max. Of course one can just delete the extra ones manually but when you're making loads it can fast become a hassle. My idea was that it could continue to work as it does now by default for people who are happy with that, but there could be an option in settings to specify a (max) number of definitions copied for users like me who want/need that functionality.

As for forms, I'm glad to see we're on the same page. I agree that the most common form alone should probably be copied by default.

As for the two examples you gave with proposed changes applied, in my personal use case the first example would definitely be an improvement because only one form is present, but it wouldn't be ideal since it would still involve having to delete extra definitions. It does look cleaner with the line breaks, but in certain SRS apps that doesn't actually make a difference because they don't recognize the breaks. And honestly, even in the ones that do, the task of deleting them was the bigger problem for me; recognizing the separation between the defintions was never really an issue (for me) even without the line breaks.

Now simple mode...that sounds like an excellent idea! If it's not possible to have an option to specify how many definitons, this would be a great alternative. And I guess if you went this route you wouldn't have to change the way the default mode works now at all. (Personally I'd still want simple mode to include the kana tab though. I'm not sure if you left that out just for the sake of the examples and because I only mentioned forms & defintions, or if you intended that the simple mode wouldn't include it.)

**Comment by birtles** [comment]
So it sounds like what we have so far is:

1. We could possibly only include the top readings when there is a clear choice (e.g. only show the starred readings, presuming we matched on one of them--or even if we haven't matched on them?).
2. Adding line breaks between the numbered senses in the "entry" copy format wouldn't hurt. (Presumably anyone who has a semi-automated workflow based on this feature is using the tab-delimited version?)
3. Adding a simplified version that only includes the first sense and drops part-of-speech information might be worthwhile.

Regarding (3) we'd need to decide:

- Do we make this a setting you enable in the options page so that it applies to both "copy as entry" and "copy as tab-separated"?
- Or do we simply make it another format to choose when you go to copy?

I guess it really needs to be the former since (a) there's no room for another format, and (b) we'd need to decide if the extra format was tab-delimited or not.

Assuming we do (3), I wonder if (1) and (2) are too risky. I wouldn't want to break anyone's workflow here.

Yes, I simply forgot to include the kana! That wasn't intentional.

**Comment by demilade5** [comment]
> Do we make this a setting you enable in the options page so that it applies to both "copy as entry" and "copy as tab-separated"?

Sounds good to me!

> Assuming we do (3), I wonder if (1) and (2) are too risky. I wouldn't want to break anyone's workflow here.

As far as (2) goes, like you said it couldn't hurt. It does look cleaner that way. I can't imagine anyone complaining about that.

As for (1), you're talking about the idea of also returning only one form in the normal, non-simplified mode right? If you're worried that could be risky, perhaps it could continue to work as it does now by default but be given a toggle in the options page? (I don't really know anything about programming so I'm not sure what is and isn't within the realms of possibility.)

**Comment by birtles** [comment]
Ok great. So it sounds like we have two actions here:

1. Add an option for simple mode that: (a) only shows top forms, (b) only includes the first sense, and (c) drops part-of-speech information from the copied information.

2. Make the regular "copy as entry" in non-simplified mode add newlines between senses and between the headwords and the senses.

> As for (1), you're talking about the idea of also returning only one form in the normal, non-simplified mode right?

Right.

> If you're worried that could be risky, perhaps it could continue to work as it does now by default but be given a toggle in the options page? (I don't really know anything about programming so I'm not sure what is and isn't within the realms of possibility.)

All of this is fairly easy to implement. However, I try to limit the number of options we add because:

1. It can be overwhelming for users
2. The design philosophy of 10ten is to try to be simple
3. The more options we have the harder it is to add new features because we have to test them with all the combinations of options

For the clipboard feature, however, it's pretty separate to the rest of the extension so (3) isn't too much of a concern in this case. I think adding the "simple mode" would be fine but I might not add the "limit headwords" option just yet.

**Comment by demilade5** [comment]
Sounds good. I look forward to the changes that you're ready to implement, especially simple mode. 

Personally I would support the idea of copying only the top or most common form in the default mode too. Given what you said about design philosophy and testing issues, I understand your reluctance and why that route might be preferable to creating more options. 

I feel like with cases like 落ち着く where you end up copying 5 forms, the additional ones aren't even useful anyway. A lot of the time they're rarely used or out of use altogether. Or they're just plain redundant as far as making flash cards is concerned because they're forms you'll recognize just by knowing the common one. So it would make sense that the default c+t behavior would be to ignore them. I feel like having to copy and paste an additional form manually in the cases where you _do_ want one, makes more sense than having to delete surplus forms in the cases where you don't want them, if only because the latter scenario is probably a lot more common in the majority of use cases.

**Comment by birtles** [comment]
I started working on this today (sorry for the delay). Regarding adding line breaks, I wonder what we should do for entries with a single sense?

i.e. should it be:

```
日本語 [にほんご, にっぽんご] (n) Japanese (language)
```

or

```
日本語 [にほんご, にっぽんご]
(n) Japanese (language)
```

The latter is more consistent but it seems a little unnecessary?

Next step is to try to be more aggressive about trimming headwords, then add an option to simplify the result further.

**Comment by birtles** [comment]
@demilade5 I've implemented this over in #1616 but have yet to merge/release it. Does this look about right?

https://github.com/birchill/10ten-ja-reader/assets/1232595/67046ea4-69f1-49b9-99bc-05b084c596b4
